cpp/src/arrow/ipc/ipc-json-simple-test.cc

```diff
@@ -0,0 +1,615 @@
+  {
+    std::unique_ptr<ArrayBuilder> builder;
+    ASSERT_OK(MakeBuilder(pool, type, &builder));
+    auto& list_builder = checked_cast<ListBuilder&>(*builder);
+    ASSERT_OK(list_builder.AppendNull());
+    ASSERT_OK(list_builder.Append());
+    ASSERT_OK(list_builder.AppendNull());
+    ASSERT_OK(list_builder.Finish(&expected));
+  }
+  AssertArraysEqual(*expected, *actual);
+}
+
+TEST(TestList, IntegerListList) {
+  auto pool = default_memory_pool();
+  std::shared_ptr<DataType> type = list(list(uint8()));
+  std::shared_ptr<Array> offsets, values, nested, expected, actual;
+
+  ASSERT_OK(ArrayFromJSON(type, "[[[4], [5, 6]], [[7, 8, 9]]]", &actual));
+  ASSERT_OK(ValidateArray(*actual));
+  ArrayFromVector<Int32Type>({0, 1, 3, 6}, &offsets);
+  ArrayFromVector<UInt8Type>({4, 5, 6, 7, 8, 9}, &values);
+  ASSERT_OK(ListArray::FromArrays(*offsets, *values, pool, &nested));
+  ArrayFromVector<Int32Type>({0, 2, 3}, &offsets);
+  ASSERT_OK(ListArray::FromArrays(*offsets, *nested, pool, &expected));
+  ASSERT_EQ(actual->length(), 2);
+  AssertArraysEqual(*expected, *actual);
+
+  ASSERT_OK(ArrayFromJSON(type, "[[], [[]], [[4], [], [5, 6]], [[7, 8, 9]]]", &actual));
+  ASSERT_OK(ValidateArray(*actual));
+  ArrayFromVector<Int32Type>({0, 0, 1, 1, 3, 6}, &offsets);
+  ArrayFromVector<UInt8Type>({4, 5, 6, 7, 8, 9}, &values);
+  ASSERT_OK(ListArray::FromArrays(*offsets, *values, pool, &nested));
+  ArrayFromVector<Int32Type>({0, 0, 1, 4, 5}, &offsets);
+  ASSERT_OK(ListArray::FromArrays(*offsets, *nested, pool, &expected));
+  ASSERT_EQ(actual->length(), 4);
+  AssertArraysEqual(*expected, *actual);
+
+  ASSERT_OK(ArrayFromJSON(type, "[null, [null], [[null]]]", &actual));
+  ASSERT_OK(ValidateArray(*actual));
+  {
+    std::unique_ptr<ArrayBuilder> builder;
+    ASSERT_OK(MakeBuilder(pool, type, &builder));
+    auto& list_builder = checked_cast<ListBuilder&>(*builder);
+    auto& child_builder = checked_cast<ListBuilder&>(*list_builder.value_builder());
+    ASSERT_OK(list_builder.AppendNull());
+    ASSERT_OK(list_builder.Append());
+    ASSERT_OK(child_builder.AppendNull());
+    ASSERT_OK(list_builder.Append());
+    ASSERT_OK(child_builder.Append());
+    ASSERT_OK(list_builder.Finish(&expected));
+  }
+}
+
+TEST(TestStruct, SimpleStruct) {
+  auto field_a = field("a", int8());
+  auto field_b = field("b", boolean());
+  std::shared_ptr<DataType> type = struct_({field_a, field_b});
+  std::shared_ptr<Array> a, b, expected, actual;
+  std::shared_ptr<Buffer> null_bitmap;
+  std::vector<bool> is_valid;
+  std::vector<std::shared_ptr<Array>> children;
+
+  ASSERT_OK(ArrayFromJSON(type, "[]", &actual));
+  ASSERT_OK(ValidateArray(*actual));
+  ArrayFromVector<Int8Type>({}, &a);
+  ArrayFromVector<BooleanType, bool>({}, &b);
+  children.assign({a, b});
+  expected = std::make_shared<StructArray>(type, 0, children);
+  AssertArraysEqual(*expected, *actual);
+
+  ASSERT_OK(ArrayFromJSON(type, "[[5, true], [6, false]]", &actual));
```

Comment:
While verbose, it would be natural to see and/or expect `[{"a": 5, "b": true}, {"b": false, "a": 6}]`. Such that you can re-order fields in the json.

Reply:
Ok, I can add this.